evalml/pipelines/component_graph.py

```diff
@@ -0,0 +1,366 @@
+                component_dict[component_name].append(f"{previous_component}.x")
+            previous_component = component_name
+        return cls(component_dict, random_state=random_state)
+
+    def instantiate(self, parameters):
+        """Instantiates all uninstantiated components within the graph using the given parameters. An error will be
+        raised if a component is already instantiated but the parameters dict contains arguments for that component.
+
+        Arguments:
+            parameters (dict): Dictionary with component names as keys and dictionary of that component's parameters as values.
+                               An empty dictionary {} implies using all default values for component parameters.
+        """
+        if self._is_instantiated:
+            raise ValueError(f"Cannot reinstantiate a component graph that was previously instantiated")
+
+        self._is_instantiated = True
+        for component_name, component_class in self.component_instances.items():
+            component_parameters = parameters.get(component_name, {})
+
+            try:
+                new_component = component_class(**component_parameters, random_state=self.random_state)
+            except (ValueError, TypeError) as e:
+                err = "Error received when instantiating component {} with the following arguments {}".format(component_name, component_parameters)
+                raise ValueError(err) from e
+
+            self.component_instances[component_name] = new_component
+        return self
+
+    def fit(self, X, y):
+        """Fit each component in the graph
+
+        Arguments:
+            X (pd.DataFrame): The input training data of shape [n_samples, n_features]
+            y (pd.Series): The target training data of length [n_samples]
+        """
+        self._compute_features(self.compute_order, X, y, fit=True)
+        return self
+
+    def fit_features(self, X, y):
+        """ Fit all components save the final one, usually an estimator. Note that this is
+        only useful where the final component has only one parent, otherwise there will be
+        information missing.
+
+        Arguments:
+            X (pd.DataFrame): The input training data of shape [n_samples, n_features]
+            y (pd.Series): The target training data of length [n_samples]
+        """
+        return self._compute_features(self.compute_order[:-1], X, y, fit=True)
+
+    def predict(self, X):
+        """Make predictions using selected features.
+
+        Arguments:
+            X (pd.DataFrame): Data of shape [n_samples, n_features]
+
+        Returns:
+            pd.Series: Predicted values.
+        """
+        return self._compute_features(self.compute_order, X)
+
+    def compute_final_component_features(self, X, y=None):
+        """ Transform all components save the final one, and gathers the data from any number of parents
+        to get all the information that should be fed to the final component
```

Comment:
I just replaced `transform_features` with this method. `transform_features` returned the output of the penultimate component, which was not useful for nonlinear pipelines where the final component has multiple parents. I was running into issues trying to use this within `PipelineBase`'s `compute_estimator_features`.

This new method should gather and combine the inputs of any/all parents of the final component, specifically to allow for this code to work:
```
X_t = compute_final_component_features(X, y)
classification_estimator.predict_proba(X_t)
```
where the estimator has multiple parents

Reply:
this looks good to me! Awesome that it supports both `Estimator` and `Transformer` outputs~

Reply:
@eccabay I'd like to propose a simplification:
* Update `_compute_features` to simply compute and return the `output_cache`, and delete the code at the end which special-cases the final component in the list.
* `fit` and `fit_features` can just ignore the output as they already do:
```python
self._compute_features(...)
return self
```
* `predict` can access the appropriate output:
```python
return self._compute_features(self.compute_order, X, y=y, fit=False)[self.compute_order[-1]]
```
* `compute_final_component_features` can do similarly:
```python
component_outputs = self._compute_features(self.compute_order, X, y=y, fit=False)
final_component_inputs = []
for parent in self.get_parents(self.compute_order[-1]):
    if parent[-2:] == '.x' or parent[-2] == '.y':
        parent = parent[:-2]
    parent_output = component_outputs[parent]
    if isinstance(parent_output, pd.Series):
        parent_output = pd.DataFrame(parent_output, columns=[parent])
    final_component_inputs.append(parent_output)
return X_t = pd.concat(final_component_inputs, axis=1)
```

The main advantage of doing this is code simplification. I think we should aim to have one implementation of pipeline evaluation (`_compute_features`) instead of two as we do now (`_compute_features` and `compute_final_component_features`), and to have all the other methods use that implementation :) It'll help keep the code simple and avoid hairy bugs arising from slight discrepancies in the separate code paths.

Reply:
@dsherry this is great, thanks for the suggestion! I was looking for a way to simplify as I was writing the method but couldn't find anything better, so I appreciate this a lot haha